src/adapter/bridges/ZKSyncBridge.ts

```diff
@@ -0,0 +1,161 @@
+import {
+  Contract,
+  BigNumber,
+  paginatedEventQuery,
+  Signer,
+  EventSearchConfig,
+  Provider,
+  spreadEventWithBlockNumber,
+  BigNumberish,
+} from "../../utils";
+import { CONTRACT_ADDRESSES } from "../../common";
+import { SortableEvent } from "../../interfaces";
+import { isDefined } from "../../utils/TypeGuards";
+import { BridgeTransactionDetails, BaseBridgeAdapter, BridgeEvents } from "./BaseBridgeAdapter";
+import { Event } from "ethers";
+import * as zksync from "zksync-web3";
+import { gasPriceOracle } from "@across-protocol/sdk";
+import { zkSync as zkSyncUtils } from "../../utils/chains";
+
+/* For both the canonical bridge (this bridge) and the ZkSync Weth
+ * bridge, we need to assume that the l1 and l2 signers contain
+ * associated providers, since we need to get information about
+ * addresses and gas prices (this is also why `constructL1toL2Txn`
+ * is an async fn).
+ */
+export class ZKSyncBridge extends BaseBridgeAdapter {
+  private readonly l1Bridge: Contract;
+  private readonly l2Bridge: Contract;
+  private readonly zkSyncMailbox: Contract;
+
+  private readonly gasPerPubdataLimit = zksync.utils.REQUIRED_L1_TO_L2_GAS_PER_PUBDATA_LIMIT;
+  private readonly l2GasLimit = 2_000_000; // We should dynamically define this.
+
+  constructor(l2chainId: number, hubChainId: number, l1Signer: Signer, l2SignerOrProvider: Signer | Provider) {
+    super(l2chainId, hubChainId, l1Signer, l2SignerOrProvider, [
+      CONTRACT_ADDRESSES[hubChainId].zkSyncDefaultErc20Bridge.address,
+    ]);
+
+    const { address: l1Address, abi: l1Abi } = CONTRACT_ADDRESSES[hubChainId].zkSyncDefaultErc20Bridge;
+    this.l1Bridge = new Contract(l1Address, l1Abi, l1Signer);
+
+    const { address: l2Address, abi: l2Abi } = CONTRACT_ADDRESSES[l2chainId].zkSyncDefaultErc20Bridge;
+    this.l2Bridge = new Contract(l2Address, l2Abi, l2SignerOrProvider);
+
+    const { address: mailboxAddress, abi: mailboxAbi } = CONTRACT_ADDRESSES[hubChainId].zkSyncMailbox;
+    this.zkSyncMailbox = new Contract(mailboxAddress, mailboxAbi, l1Signer);
+  }
+
+  async constructL1ToL2Txn(
+    toAddress: string,
+    l1Token: string,
+    l2Token: string,
+    amount: BigNumber
+  ): Promise<BridgeTransactionDetails> {
+    const l1Provider = this.l1Bridge.provider;
+    const l2Provider = this.l2Bridge.provider;
+
+    let zkProvider;
+    try {
+      zkProvider = zkSyncUtils.convertEthersRPCToZKSyncRPC(l2Provider);
+    } catch (error) {
+      // We do not have a logger in this class
+    }
+    // If zkSync provider can't be created for some reason, default to a very conservative 2mil L2 gas limit
+    // which should be sufficient for this transaction.
+    const l2GasLimit = isDefined(zkProvider)
+      ? await zksync.utils.estimateDefaultBridgeDepositL2Gas(
+          l1Provider,
+          zkProvider,
+          l1Token,
+          amount,
+          toAddress,
+          toAddress,
+          this.gasPerPubdataLimit
+        )
+      : 2_000_000;
```

Comment:
This is a pre-existing issue, but returning a number here results in l2GasLimit type awkwardness (union number | BigNumber). We can just assign a BN here to fix that.

```suggestion
      : BigNumber.from(2_000_000);
```

Reply:
Same as below: https://github.com/across-protocol/relayer/pull/1524/commits/5aacd351220a191e4ba57c07e729bccbbdb36020